modules/phase_field/src/materials/DerivativeParsedMaterialHelper.C

```diff
@@ -9,139 +9,106 @@
-    ParsedMaterialHelper<DerivativeFunctionMaterialBase>(name, parameters, map_mode)
+    ParsedMaterialHelper(name, parameters, map_mode),
+    //_derivative_order(getParam<unsigned int>("derivative_order"))
+    _derivative_order(isParamValid("third_derivatives") ? (getParam<bool>("third_derivatives") ? 3 : 2) : getParam<unsigned int>("derivative_order"))
 {
 }
 
 DerivativeParsedMaterialHelper::~DerivativeParsedMaterialHelper()
 {
-  for (unsigned int i = 0; i < _nargs; ++i)
-  {
-    delete _func_dF[i];
-
-    for (unsigned int j = i; j < _nargs; ++j)
-    {
-      delete _func_d2F[i][j];
-
-      if (_third_derivatives)
-        for (unsigned int k = j; k < _nargs; ++k)
-          delete _func_d3F[i][j][k];
-    }
-  }
+  for (unsigned int i = 0; i < _derivatives.size(); ++i)
+    delete _derivatives[i].second;
 }
 
 void DerivativeParsedMaterialHelper::functionsPostParse()
 {
-  functionsDerivative();
-  functionsOptimize();
-}
-
-void
-DerivativeParsedMaterialHelper::functionsDerivative()
-{
-  unsigned int i, j, k;
+  // optimize base function
+  ParsedMaterialHelper::functionsOptimize();
 
-  // first derivatives
-  _func_dF.resize(_nargs);
-  _func_d2F.resize(_nargs);
-  _func_d3F.resize(_nargs);
-  for (i = 0; i < _nargs; ++i)
-  {
-    _func_dF[i] = new ADFunction(*_func_F);
-    if (_func_dF[i]->AutoDiff(_arg_names[i]) != -1)
-      mooseError("Failed to take first derivative w.r.t. "
-                 << _arg_names[i]);
-
-    // second derivatives
-    _func_d2F[i].resize(_nargs);
-    _func_d3F[i].resize(_nargs);
-    for (j = i; j < _nargs; ++j)
-    {
-      _func_d2F[i][j] = new ADFunction(*_func_dF[i]);
-      if (_func_d2F[i][j]->AutoDiff(_arg_names[j]) != -1)
-        mooseError("Failed to take second derivative w.r.t. "
-                   << _arg_names[i] << " and " << _arg_names[j]);
-
-      // third derivatives
-      if (_third_derivatives)
-      {
-        _func_d3F[i][j].resize(_nargs);
-        for (k = j; k < _nargs; ++k)
-        {
-          _func_d3F[i][j][k] = new ADFunction(*_func_d2F[i][j]);
-          if (_func_d3F[i][j][k]->AutoDiff(_arg_names[k]) != -1)
-            mooseError("Failed to take third derivative w.r.t. "
-                       << _arg_names[i] << ", " << _arg_names[j] << ", and " << _arg_names[k]);
-        }
-      }
-    }
-  }
+  // generate derivatives
+  assembleDerivatives();
 }
 
+/**
+ * Peform a breadth first construction of all requeste derivatives.
+ */
 void
-DerivativeParsedMaterialHelper::functionsOptimize()
+DerivativeParsedMaterialHelper::assembleDerivatives()
 {
-  unsigned int i, j, k;
+  // need to check for zero derivatives here, otherwise at least one order is generated
+  if (_derivative_order < 1) return;
 
-  // base function
-  ParsedMaterialHelper<DerivativeFunctionMaterialBase>::functionsOptimize();
+  // set up deque
+  std::deque<QueueItem> queue;
+  queue.push_back(QueueItem(_func_F));
```

Comment:
Yeah, I never knew that. I really though all C++ containers had iterators.
We could change the other "deque" in MOOSE to a queue if we wanted since we
never iterate over it, but it's a spurious change.


On Fri, May 1, 2015 at 9:34 AM Daniel Schwen <notifications@github.com>
wrote:

> In modules/phase_field/src/materials/DerivativeParsedMaterialHelper.C
> <https://github.com/idaholab/moose/pull/5014#discussion_r29508631>:
>
> >
> > -  // base function
> > -  ParsedMaterialHelper<DerivativeFunctionMaterialBase>::functionsOptimize();
> > +  // set up deque
> > +  std::deque<QueueItem> queue;
> > +  queue.push_back(QueueItem(_func_F));
>
> Just for the record if anyone should read this: std::queue does not
> provide an iterator over the items in the underlying container, and
> iteration is necessary for the algorithm that will be part of the follow on
> PR. So this has to be a std::deque.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5014/files#r29508631>.
>